compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirNotImplementedOverrideChecker.kt

```diff
@@ -190,6 +133,90 @@ object FirNotImplementedOverrideChecker : FirClassChecker() {
         }
     }
 
+    fun collectSymbols(
+        firClass: FirClass<*>,
+        sessionHolder: SessionHolder,
+        collect: (
+            symbol: FirCallableSymbol<*>,
+            isNotImplemented: Boolean,
+            isAmbiguouslyImplemented: Boolean,
+            isVisible: Boolean,
+            isOverridable: Boolean
+        ) -> Unit
+    ) {
+        val classScope = firClass.unsubstitutedScope(sessionHolder.session, sessionHolder.scopeSession, withForcedTypeCalculator = false)
+        val classPackage = firClass.symbol.classId.packageFqName
+
+        fun FirCallableMemberDeclaration<*>.isVisible(): Boolean {
+            if (visibility == Visibilities.Private ||
+                !visibility.visibleFromPackage(classPackage, symbol.callableId.packageName)
+            ) return false
+            if (visibility == Visibilities.Internal &&
+                session !== firClass.session
+            ) return false
+            return true
+        }
+
+        fun FirCallableSymbol<*>.implementationCount(): Int {
+            val fir = this.fir as FirCallableMemberDeclaration<*>
+            val containingClass = fir.getContainingClass(sessionHolder)
+            if (this is FirIntersectionCallableSymbol &&
+                containingClass === firClass
+            ) {
+                val count = intersections.count { (it.fir as FirCallableMemberDeclaration).modality != Modality.ABSTRACT }
+                if (count > 1) return count
+            }
+            if (fir is FirSimpleFunction) {
+                if (firClass is FirRegularClass && firClass.isData && fir.matchesDataClassSyntheticMemberSignatures) {
+                    return 1
+                }
+                // TODO: suspend function overridden by a Java class in the middle is not properly regarded as an override
+                if (fir.isSuspend) {
+                    return 1
+                }
+            }
+            if (!fir.isAbstract) return 1
+            if (containingClass === firClass && fir.origin == FirDeclarationOrigin.Source) return 1
+            if (containingClass is FirRegularClass && containingClass.isExpect) return 1
+            return 0
```

Comment:
All of these counts don't make sense to me, and I found that this is because somewhat incompatible notions,  `shouldBeImplemented` and `intersections.count`, are merged. I wonder if you can stick to the original notions.

Reply:
I actually find it cleaner this way because now the return type `Int` makes it impossible at type level to have both `notImplemented` and `isAmbiguouslyInherited` to be true. Separating the logic would put the burden on the programmer to enforce this. How about renaming the function to `countAvailableImplementationInheritedFromSuperTypes`?

Reply:
Actually let me use a enum instead.